src/routes/versions.rs

```diff
@@ -518,6 +581,100 @@ pub async fn get_version_from_hash(
     }
 }
 
+#[derive(Serialize, Deserialize)]
+pub struct DownloadRedirect {
+    pub url: String,
+}
+
+// under /api/v1/version_file/{hash}
+#[post("{version_id}")]
+pub async fn download_version(
+    req: HttpRequest,
+    info: web::Path<(String,)>,
+    pool: web::Data<PgPool>,
+    algorithm: web::Query<Algorithm>,
+) -> Result<HttpResponse, ApiError> {
+    let hash = info.into_inner().0;
+
+    let result = sqlx::query!(
+        "
+        SELECT f.url url, f.id id, f.version_id version_id, v.mod_id mod_id FROM files f
+        INNER JOIN versions v ON v.id = f.version_id
+        INNER JOIN hashes ON hash = $1 AND algorithm = $2
+        ",
+        hash.as_bytes(),
+        algorithm.algorithm
+    )
+    .fetch_optional(&**pool)
+    .await
+    .map_err(|e| ApiError::DatabaseError(e.into()))?;
+
+    if let Some(id) = result {
+        let download_exists = sqlx::query!(
+            "SELECT EXISTS(SELECT 1 FROM downloads WHERE version_id = $1 AND date > (CURRENT_DATE - INTERVAL '30 minutes ago'))",
+            id.version_id,
+        )
+            .fetch_one(&**pool)
+            .await
+            .map_err(|e| ApiError::DatabaseError(e.into()))?
+            .exists.unwrap_or(false);
+
+        let real_ip = req.connection_info();
+        if !download_exists {
+            sqlx::query!(
+                "
+                INSERT INTO downloads (
+                    version_id, identifier
+                )
+                VALUES (
+                    $1, $2
+                )
+                ",
+                id.version_id,
+                real_ip
+                    .realip_remote_addr()
+                    .ok_or_else(|| ApiError::InvalidInputError(
+                        "You may only record a download for a mod with a normal client!"
+                            .to_string()
+                    ))?
```

Comment:
Isn't this supposed to be the sha-1 hash of the IP address?  Also, this should not throw any errors if it can't log it - it should still redirect and let the user download it.